File: Sources/DoNotTypeCore/FileTranscriber.swift
import Foundation

/// Transcribes a recording that already exists, rather than one being spoken right now.
///
/// ## Why this is a separate type from the live path
///
/// Live dictation is a latency problem: the user is standing there, the audio is already in the
/// right format, and every decision — pre-upload, hedging, chunk concurrency — exists to shorten
/// the wait. A file on disk is a *throughput* problem. It may be forty minutes long, in a format
/// nothing downstream understands, and nobody is waiting on the first word.
///
/// The two share everything that matters (`TranscriptionService`, the prompt, the chunker, the
/// providers) and differ in what wraps them, which is what this type is: decode, transcribe, and
/// then optionally run the second stage the mode asks for.
///
/// ## The invariant
///
/// The verbatim transcript is produced first and returned alongside whatever was derived from it,
/// always — including for summaries, where the derived text is by definition missing most of what
/// was said. A caller that only wants the summary can ignore `verbatim`; a caller that discards it
/// has made that choice explicitly, which is the whole difference this project is arguing for.
public struct FileTranscriber: Sendable {
    /// What the caller can show while this runs. A forty-minute file is not a spinner.
    public enum Progress: Sendable, Equatable {
        case decoding(String)
        case transcribing(done: Int, of: Int)
        case deriving(TranscriptMode)
    }

    public struct Outcome: Sendable {
        public var sourceURL: URL
        /// Word for word, at the requested fidelity. Always present.
        public var verbatim: String
        /// What the mode produced. Identical to `verbatim` for `.verbatim`.
        public var delivered: String
        public var mode: TranscriptMode
        public var fidelity: Fidelity
        public var language: String
        public var usage: TokenUsage
        /// How many requests the audio was split across.
        public var chunkCount: Int
        public var durationSeconds: Double?
        public var decodeSeconds: Double
        public var transcriptionSeconds: Double
        /// Time in the second stage, when there was one.
        public var secondStageSeconds: Double?
        public var provider: String
        public var model: String
        /// The backend that ran the second stage, when it was not the transcription one.
        public var secondStageProvider: String?

        public var totalSeconds: Double {
            decodeSeconds + transcriptionSeconds + (secondStageSeconds ?? 0)
        }

        /// How much faster than real time the whole thing ran. Nil when the duration is unknown.
        public var realtimeFactor: Double? {
            guard let durationSeconds, durationSeconds > 0 else { return nil }
            return totalSeconds / durationSeconds
        }

        /// A history row for this file, so an offline transcription is searchable next to the
        /// dictations — and so `dnt` and the app agree on what one looks like.
        public func historyRecord() -> DictationRecord {
            DictationRecord(
                status: .completed,
                text: verbatim,
                styledText: mode == .verbatim ? nil : delivered,
                style: mode.rewriteStyle,
                provider: provider,
                model: model,
                fidelity: fidelity,
                durationSeconds: durationSeconds ?? 0,
                latencySeconds: totalSeconds,
                requestSeconds: transcriptionSeconds,
                rewriteSeconds: secondStageSeconds,
                chunkCount: chunkCount,
                usage: usage,
                mode: mode,
                sourceFileName: sourceURL.lastPathComponent)
        }
    }

    /// Transcribes the audio.
    public var service: TranscriptionService
    /// Runs the second stage, when the transcription backend cannot.
    ///
    /// This is what makes `--mode summary --provider xai` work at all: xAI's endpoint is a
    /// recogniser and has no text input, so summarising through it is not a slow path, it is an
    /// impossible one. Pointing the second stage at a model keeps the fast recogniser for the part
    /// it is good at.
    public var secondStage: TranscriptionService?
    public var prompt: PromptBuilder
    public var fidelity: Fidelity

    private let log = Log("file")

    public init(
        service: TranscriptionService,
        prompt: PromptBuilder,
        fidelity: Fidelity = .default,
        secondStage: TranscriptionService? = nil
    ) {
        self.service = service
        self.prompt = prompt
        self.fidelity = fidelity
        self.secondStage = secondStage
    }

    /// The backend that will run the second stage, or nil when nothing available can.
    ///
    /// A recogniser has no text channel, so this is a capability question rather than a preference:
    /// `GroundingSupport.readsSystemInstruction` is exactly "this backend is a language model".
    var textCapableService: TranscriptionService? {
        if let secondStage, secondStage.grounding.readsSystemInstruction { return secondStage }
        if service.grounding.readsSystemInstruction { return service }
        return nil
    }

    /// Whether this configuration can run the given mode at all, checked before any audio is sent.
    ///
    /// Asked up front deliberately. Discovering that a summary is impossible *after* billing forty
    /// minutes of audio would be an expensive way to learn it.
    public func supports(_ mode: TranscriptMode) -> Bool {
        !mode.needsSecondPass || textCapableService != nil
    }

    public func transcribe(
        fileAt url: URL,
        mode: TranscriptMode = .verbatim,
        context: ScreenContext? = nil,
        verifyNumbers: Bool = false,
        attempts: Int = 3,
        maxConcurrent: Int = 3,
        onProgress: (@Sendable (Progress) -> Void)? = nil
    ) async throws -> Outcome {
        guard supports(mode) else {
            throw ProviderError.audioRequired(provider: service.provider.name)
        }

        let name = url.lastPathComponent
        log.info(
            "transcribing file",
            [
                "file": name, "mode": mode.rawValue, "provider": service.provider.name,
                "model": service.model, "fidelity": fidelity.rawValue,
            ])

        onProgress?(.decoding(name))
        let decodeStart = Date()
        let audio = try AudioDecoder.load(url)
        let decodeSeconds = Date().timeIntervalSince(decodeStart)

        let transcribeStart = Date()
        let result = try await service.transcribeLong(
            audio: audio, context: context, attempts: attempts, maxConcurrent: maxConcurrent,
            verifyNumbers: verifyNumbers,
            onProgress: { done, total in onProgress?(.transcribing(done: done, of: total)) })
        let transcriptionSeconds = Date().timeIntervalSince(transcribeStart)

        let verbatim = result.transcript.transcript.trimmed
        log.info(
            "transcribed file",
            [
                "file": name, "chars": "\(verbatim.count)", "chunks": "\(result.chunkCount)",
                "ms": LogClock.ms(transcriptionSeconds),
            ])
        log.content("transcript", verbatim, level: .trace)

        var outcome = Outcome(
            sourceURL: url,
            verbatim: verbatim,
            delivered: verbatim,
            mode: mode,
            fidelity: fidelity,
            language: result.transcript.language,
            usage: result.usage,
            chunkCount: result.chunkCount,
            durationSeconds: audio.durationSeconds,
            decodeSeconds: decodeSeconds,
            transcriptionSeconds: transcriptionSeconds,
            provider: service.provider.name,
            model: service.model)

        // An empty transcript means silence, and there is nothing to rewrite or summarise. Running
        // the second stage anyway would ask a model to write prose from nothing, which is the one
        // way this pipeline could invent words outright.
        guard mode.needsSecondPass, !verbatim.isEmpty else { return outcome }

        guard let instruction = try prompt.secondStageInstruction(for: mode),
            let deriver = textCapableService
        else { return outcome }

        onProgress?(.deriving(mode))
        let deriveStart = Date()
        let derived = try await deriver.rewrite(verbatim, instruction: instruction)
        outcome.secondStageSeconds = Date().timeIntervalSince(deriveStart)
        outcome.delivered = derived
        outcome.secondStageProvider = deriver.provider.name == service.provider.name
            ? nil : deriver.provider.name
        log.info(
            "second stage finished",
            [
                "file": name, "mode": mode.rawValue, "chars": "\(derived.count)",
                "from": "\(verbatim.count)",
                "ms": LogClock.ms(outcome.secondStageSeconds ?? 0),
                "provider": deriver.provider.name,
            ])
        return outcome
    }
}

File: windows/DoNotType.Core/FileTranscriber.cs
namespace DoNotType.Core;

/// <summary>
/// Transcribes a recording that already exists, rather than one being spoken right now.
/// </summary>
/// <remarks>
/// <para>
/// Live dictation is a latency problem: the user is standing there, the audio is already in the
/// right format, and every decision exists to shorten the wait. A file on disk is a
/// <em>throughput</em> problem -- it may be forty minutes long, and nobody is waiting on the first
/// word.
/// </para>
/// <para>
/// The invariant is the one the whole project rests on: the verbatim transcript is produced and
/// returned alongside whatever was derived from it, always, including for summaries where the
/// derived text is by definition missing most of what was said.
/// </para>
/// </remarks>
public sealed class FileTranscriber(
    TranscriptionService service,
    PromptBuilder prompt,
    Fidelity fidelity = Fidelity.Light,
    TranscriptionService? secondStage = null)
{
    private static readonly Log Log = new("file");

    /// <summary>What the caller can show while this runs. A forty-minute file is not a spinner.</summary>
    public abstract record Progress
    {
        private Progress() { }

        public sealed record Decoding(string File) : Progress;
        public sealed record Transcribing(int Done, int Of) : Progress;
        public sealed record Deriving(TranscriptMode Mode) : Progress;
    }

    public sealed record Outcome
    {
        public required string SourcePath { get; init; }

        /// <summary>Word for word, at the requested fidelity. Always present.</summary>
        public required string Verbatim { get; init; }

        /// <summary>What the mode produced. Identical to <see cref="Verbatim"/> for verbatim.</summary>
        public required string Delivered { get; init; }

        public required TranscriptMode Mode { get; init; }
        public required Fidelity Fidelity { get; init; }
        public string Language { get; init; } = string.Empty;
        public TokenUsage Usage { get; init; } = new();
        public int ChunkCount { get; init; } = 1;
        public double DurationSeconds { get; init; }
        public double DecodeSeconds { get; init; }
        public double TranscriptionSeconds { get; init; }
        public double? SecondStageSeconds { get; init; }
        public required string Provider { get; init; }
        public required string Model { get; init; }

        /// <summary>The backend that ran the second stage, when it was not the transcription one.</summary>
        public string? SecondStageProvider { get; init; }

        public double TotalSeconds => DecodeSeconds + TranscriptionSeconds + (SecondStageSeconds ?? 0);

        /// <summary>
        /// A history row for this file, so an offline transcription is searchable next to the
        /// dictations -- and so the CLI and the app agree on what one looks like.
        /// </summary>
        public DictationRecord ToRecord() => new()
        {
            Status = DictationStatus.Completed,
            Text = Verbatim,
            StyledText = Mode is TranscriptMode.VerbatimMode ? null : Delivered,
            Mode = Mode.Id,
            SourceFileName = Path.GetFileName(SourcePath),
            Model = Model,
            Fidelity = Fidelity,
            DurationSeconds = DurationSeconds,
            LatencySeconds = TotalSeconds,
            RequestSeconds = TranscriptionSeconds,
            AudioTokens = Usage.AudioTokens,
        };
    }

    /// <summary>
    /// The backend that can run a second stage, or null when nothing available can.
    /// </summary>
    /// <remarks>
    /// A recogniser has no text channel, so this is a capability question rather than a preference.
    /// </remarks>
    private TranscriptionService? TextCapableService =>
        secondStage is not null && secondStage.Provider.Grounding is GroundingSupport.MultimodalGrounding
            ? secondStage
            : service.Provider.Grounding is GroundingSupport.MultimodalGrounding ? service : null;

    /// <summary>
    /// Whether this configuration can run the mode at all, checked before any audio is sent.
    /// Discovering that a summary is impossible after billing forty minutes would be expensive.
    /// </summary>
    public bool Supports(TranscriptMode mode) =>
        !mode.NeedsSecondPass || (TextCapableService is not null && prompt.SupportsSecondStage(mode));

    public async Task<Outcome> TranscribeAsync(
        string path,
        TranscriptMode? mode = null,
        ScreenContext? context = null,
        int attempts = 3,
        int maxConcurrent = 3,
        Action<Progress>? onProgress = null,
        CancellationToken cancellationToken = default)
    {
        mode ??= TranscriptMode.Verbatim;
        if (!Supports(mode))
        {
            throw new ProviderException(
                $"{service.Provider.Name} is a speech recognition endpoint: it transcribes audio and "
                + $"cannot do anything with text, so it cannot produce a {mode.Id}. Transcribe with "
                + "a model instead, or pair the recogniser with one for the second stage.");
        }

        var name = Path.GetFileName(path);
        Log.Info(() => "transcribing file", new Dictionary<string, string>
        {
            ["file"] = name,
            ["mode"] = mode.Id,
            ["provider"] = service.Provider.Name,
            ["model"] = service.Provider.Model,
            ["fidelity"] = fidelity.Id(),
        });

        onProgress?.Invoke(new Progress.Decoding(name));
        var decodeStart = DateTimeOffset.Now;
        var wav = AudioDecoder.Load(path);
        var decodeSeconds = (DateTimeOffset.Now - decodeStart).TotalSeconds;

        // The same gate as live dictation, for the same reason: a model handed a recording with no
        // speech in it returns a plausible sentence rather than nothing. Here it is an error rather
        // than a silent no-op, because somebody who pointed at a file and pressed go is owed an
        // answer — and "there is no speech in this recording" is a better one than a paragraph the
        // model made up.
        var activity = SpeechActivity.MeasureWav(wav);
        if (!activity.HasSpeech)
        {
            Log.Info(() => "no speech in the recording", new Dictionary<string, string>
            {
                ["file"] = name,
                ["audio"] = activity.Summary,
            });
            throw new NoSpeechException(name);
        }

        var transcribeStart = DateTimeOffset.Now;
        var result = await service.TranscribeLongAsync(
                wav, context, attempts, maxConcurrent,
                (done, of) => onProgress?.Invoke(new Progress.Transcribing(done, of)),
                cancellationToken)
            .ConfigureAwait(false);
        var transcriptionSeconds = (DateTimeOffset.Now - transcribeStart).TotalSeconds;

        var verbatim = result.Transcript.Text.Trim();
        Log.Info(() => "transcribed file", new Dictionary<string, string>
        {
            ["file"] = name,
            ["chars"] = verbatim.Length.ToString(),
            ["chunks"] = result.ChunkCount.ToString(),
            ["ms"] = ((long)(transcriptionSeconds * 1000)).ToString(),
        });
        Log.Content("transcript", () => verbatim, LogLevel.Trace);

        var outcome = new Outcome
        {
            SourcePath = path,
            Verbatim = verbatim,
            Delivered = verbatim,
            Mode = mode,
            Fidelity = fidelity,
            Language = result.Transcript.Language,
            Usage = result.Usage,
            ChunkCount = result.ChunkCount,
            DurationSeconds = AudioChunker.DurationSeconds(wav),
            DecodeSeconds = decodeSeconds,
            TranscriptionSeconds = transcriptionSeconds,
            Provider = service.Provider.Name,
            Model = service.Provider.Model,
        };

        // An empty transcript means silence, and there is nothing to rewrite or summarise. Running
        // the second stage anyway would ask a model to write prose from nothing, which is the one
        // way this pipeline could invent words outright.
        if (!mode.NeedsSecondPass || verbatim.Length == 0) return outcome;

        var instruction = prompt.SecondStageInstruction(mode);
        var deriver = TextCapableService;
        if (instruction is null || deriver is null) return outcome;

        onProgress?.Invoke(new Progress.Deriving(mode));
        var deriveStart = DateTimeOffset.Now;
        var derived = (await deriver.RewriteAsync(verbatim, instruction, cancellationToken)
            .ConfigureAwait(false)).Trim();
        var secondStageSeconds = (DateTimeOffset.Now - deriveStart).TotalSeconds;

        Log.Info(() => "second stage finished", new Dictionary<string, string>
        {
            ["file"] = name,
            ["mode"] = mode.Id,
            ["chars"] = derived.Length.ToString(),
            ["from"] = verbatim.Length.ToString(),
            ["provider"] = deriver.Provider.Name,
        });

        return outcome with
        {
            Delivered = derived.Length > 0 ? derived : verbatim,
            SecondStageSeconds = secondStageSeconds,
            SecondStageProvider =
                deriver.Provider.Name == service.Provider.Name ? null : deriver.Provider.Name,
        };
    }

    /// <summary>One output name per input, unique by construction.</summary>
    /// <remarks>
    /// `a/speech.wav` and `b/speech.mp3` both wanted to be `speech.txt`, and the second silently
    /// replaced the first — a whole transcription gone, already paid for, with "wrote speech.txt"
    /// printed twice as though both had landed. Only names that would actually collide are made
    /// ugly, because the ordinary case is one file and `speech.txt` is what anybody would expect.
    /// </remarks>
    public static IReadOnlyList<string> OutputNames(IReadOnlyList<string> paths)
    {
        var used = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var names = new List<string>(paths.Count);
        foreach (var path in paths)
        {
            var stem = Path.GetFileNameWithoutExtension(path);
            var name = stem + ".txt";
            // Keep the source extension, which is what distinguishes them and what the user typed.
            if (used.Contains(name)) name = Path.GetFileName(path) + ".txt";
            // Same name in two directories. Nothing in the name can separate those, so number them
            // in the order they were given, which is the order the "wrote ..." lines print in.
            for (var suffix = 2; used.Contains(name); suffix++) name = $"{stem}-{suffix}.txt";
            used.Add(name);
            names.Add(name);
        }
        return names;
    }
}

/// <summary>The recording contains no speech, so nothing was sent.</summary>
/// <remarks>
/// Its own type so a caller can tell "there was nothing to transcribe" from "the request failed",
/// which are different things to tell somebody and different things to retry.
/// </remarks>
public sealed class NoSpeechException(string name) : Exception(
    $"{name} has no speech in it — nothing was sent. A model given a recording with only silence "
    + "or noise in it does not reliably return nothing; it returns a plausible sentence, which is "
    + "worse than an error.");
